### PR TITLE
[codex] add automerge-only CI auto-retry

### PR DESCRIPTION
**Pull request by turadg**
## Summary
This change adds opt-in CI auto-retry behavior for PRs that have an `automerge:*` label. The goal is to avoid slowing down normal developer feedback with blanket retries, while still reducing merge friction once the author has explicitly signaled that the PR is ready to land.

The implementation introduces a dedicated workflow that watches completed runs for the monitored PR workflows and re-runs only failed jobs when the run belongs to a PR with an automerge label. For `workflow_run` events, it now fetches the full pull request metadata before checking labels, because the event payload only contains minimal PR references. It also handles the case where a workflow already failed before the label was added by scanning the latest failed runs for the PR head SHA and retrying those once the label is applied. Retries are capped at up to 3 automatic retries, which corresponds to allowing auto-retry while GitHub's 1-based `run_attempt` is less than 4.

A companion executable `.mts` helper script owns the retry policy and GitHub API calls so the workflow logic stays small and the policy is testable. The workflow now installs Node dependencies before executing the shebang-based script so `ts-blank-space/register` is available at runtime.

## User Impact
Developers working on normal PRs continue to get immediate failure feedback with no automatic retry behavior. Developers who add an `automerge:*` label get up to 3 automatic failed-job retries for monitored CI workflows, which should reduce manual reruns caused by flakes right before merge.

## Validation
I validated the helper policy with focused AVA coverage in `packages/internal/test/auto-retry-automerge.test.js`, covering automerge label detection, workflow-run eligibility, and stale-run filtering by head SHA. I also verified the `.mts` entrypoint executes through the loader path and checked the patch with `git diff --check`.

### CODE REVIEW

File: packages/internal/test/auto-retry-automerge.test.js [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+/**
+ * @param {string} expression
+ * @returns {unknown}
+ */
+const evalModuleExpression = expression => {
+  const stdout = execFileSync(
+    process.execPath,
+    [
+      '--experimental-strip-types',
```
Comment (turadg) [inline packages/internal/test/auto-retry-automerge.test.js:20]: Seeking feedback on whether to switch to ts-blank-space for this test or make it skip if Node version is too old (<22.6) to use this option. Another option is to rely on the shebang instead of importing the module. I could also make the shebang use this option instead of ts-blank-space.

File: .github/workflows/auto-retry-automerge.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-node@v4
+        with:
+          node-version: 22
+      - name: Set up Corepack Yarn
+        run: bash scripts/ensure-corepack-yarn.sh
+      - name: Install Node dependencies
+        run: yarn install --mode=skip-build --immutable
```
Comment (mhofman) [inline .github/workflows/auto-retry-automerge.yml]: A bit unfortunate we have to install all deps of the repo for this. Wondering if we need a "CI sdk independent" yarn workspace (effectively like a3p / multichain)

File: .github/workflows/auto-retry-automerge.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    types:
+      - labeled
+
+permissions:
+  actions: write
+  contents: read
+  pull-requests: read
+
+jobs:
+  retry-from-workflow-run:
```
Comment (mhofman) [inline .github/workflows/auto-retry-automerge.yml:22]: I'm confused, why do we need 2 different jobs with different conditions but what looks like the same steps?

File: packages/internal/test/auto-retry-automerge.test.js (hunk 1 above)
Comment (mhofman) [reply packages/internal/test/auto-retry-automerge.test.js:20]: skip the test for old Node.js. It only ever runs in an up to date node version

File: packages/internal/test/auto-retry-automerge.test.js [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+  t.true(
+    evalModuleExpression(
+      `mod.hasAutomergeLabel([{ name: 'automerge:rebase' }])`,
+    ),
+  );
+  t.false(
+    evalModuleExpression(
+      `mod.hasAutomergeLabel([{ name: 'bypass:automerge' }])`,
+    ),
+  );
```
Comment (mhofman) [inline packages/internal/test/auto-retry-automerge.test.js]: I would argue this one should also qualify.

File: .github/scripts/auto-retry-automerge.mts [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+const selectLatestRetryableRuns = (
+  runs: WorkflowRun[],
+  headSha: string,
+): WorkflowRun[] => {
+  const latestByWorkflow = new Map<string, WorkflowRun>();
+
+  for (const run of runs) {
+    if (!isRetryableForHeadSha(run, headSha) || !run.name) {
+      continue;
+    }
```
Comment (mhofman) [inline .github/scripts/auto-retry-automerge.mts]: I think bailing out this early will not allow clearing a run that had previously failed then succeeded.

It likely means a first iteration to reduce to the latest run, then a filter to keep only the failed workflows.

I'd like to see a test of this:
- one workflow that fails, it gets re-run
- another workflow that fails, it gets re-run
- the first workflow completes successfully
- the second workflow fails again. Re-running it should not attempt to re-run the "failed jobs" of the first workflow

File: .github/scripts/auto-retry-automerge.mts [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+  /** @type {WorkflowRun[]} */
+  const runs = [];
+
+  for (let page = 1; ; page += 1) {
+    const response = (await githubRequest(
+      'GET',
+      `/repos/${owner}/${repo}/actions/runs?event=pull_request&status=completed&head_sha=${encodeURIComponent(
+        headSha,
+      )}&per_page=100&page=${page}`,
+    )) as { workflow_runs?: WorkflowRun[] };
```
Comment (mhofman) [inline .github/scripts/auto-retry-automerge.mts]: I think listing only the completed jobs will cause us to lose knowledge of any re-run already in progress, and as such attempt re-run of a workflow that is already re-running.

This would happen for example in case of applying the label when a re-run is already in progress, or if we have multiple workflows and one completed after a re-run has already been triggered.

File: packages/internal/test/auto-retry-automerge.test.js [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    {
+      id: 11,
+      name: 'Test all Packages',
+      head_sha: headSha,
+      conclusion: 'failure',
+      event: 'pull_request',
+      run_attempt: 1,
+      pull_requests: [{ number: 1 }],
+      created_at: '2026-03-06T10:05:00Z',
+    },
```
Comment (mhofman) [inline packages/internal/test/auto-retry-automerge.test.js]: How can we have 2 different runs with the same attempt number for the same workflow?

**Review by mhofman — CHANGES_REQUESTED** [review]
There doesn't seem to be sufficient edge case checks of the retriable runs logic, in particular with runs becoming green after failure, or runs being in progress when triggering the script.

File: packages/internal/test/auto-retry-automerge.test.js (hunk 7 above)
Comment (turadg) [reply packages/internal/test/auto-retry-automerge.test.js]: Separate workflow runs for the same SHA can happen from multiple PR events, and each distinct workflow run starts at run_attempt = 1.

File: packages/internal/test/auto-retry-automerge.test.js (hunk 4 above)
Comment (turadg) [reply packages/internal/test/auto-retry-automerge.test.js]: Ok, change of spec but I'll include it.

File: .github/workflows/auto-retry-automerge.yml (hunk 3 above)
Comment (turadg) [reply .github/workflows/auto-retry-automerge.yml:22]: Codex:

Because the two triggers have different payloads and different retry scopes, even though they currently happen to share setup.

`workflow_run` is reactive to one specific failed workflow completion. That path should inspect that single run and decide whether to rerun its failed jobs.

`pull_request:labeled` is recovery logic. When someone adds `automerge:*` after failures already happened, there is no single failed run in the event payload, so that path has to scan prior runs for the PR head SHA and retry the relevant ones.

So the jobs are semantically different:
- `retry-from-workflow-run`: handle one newly failed monitored run
- `retry-from-label`: backfill retries for already-failed monitored runs

The duplication is mostly just environment setup. If we want, we can factor that out:
- simplest: keep two jobs, extract the shared setup into a composite action or reusable workflow
- cleaner in this case: collapse to one job and let the script branch on `GITHUB_EVENT_NAME`


----
Would you prefer one job plus the script branching?

File: .github/workflows/auto-retry-automerge.yml (hunk 2 above)
Comment (turadg) [reply .github/workflows/auto-retry-automerge.yml]: Good point. That helps resolve the ts-blank-space decision.

File: packages/internal/test/auto-retry-automerge.test.js (hunk 7 above)
Comment (mhofman) [reply packages/internal/test/auto-retry-automerge.test.js]: Oh right, labelling is still considered a `pull_request` event.

File: .github/workflows/auto-retry-automerge.yml (hunk 3 above)
Comment (mhofman) [reply .github/workflows/auto-retry-automerge.yml:22]: I'm ok keeping as-is too. I was mostly concerned about unintended divergence under maintenance.
